File: adventofcode/2019/d2/d2.cpp
#include <fstream>
#include <iostream>
#include <vector>
using namespace std;

int run(vector<int> program) {
  vector<int> arr = program;
  auto i = 0;
  while (true) {
    auto op = arr[i];
    if (op == 99) {
      return arr[0];
    }
    auto a = arr[i + 1];
    auto b = arr[i + 2];
    auto to = arr[i + 3];
    if (op == 1) {
      arr[to] = arr[a] + arr[b];
    } else if (op == 2) {
      arr[to] = arr[a] * arr[b];
    }
    i += 4;
  }
}

int main() {
  ifstream ifile("./input.txt");
  vector<int> program;
  while (ifile) {
    char c;
    int i;
    ifile >> i >> c;
    program.push_back(i);
  }

  vector<int> p1 = program;
  p1[1] = 12;
  p1[2] = 2;
  cout << run(p1) << "\n";

  for (int x = 0; x < 100; ++x) {
    for (int y = 0; y < 100; ++y) {
      vector<int> p2 = program;
      p2[1] = x;
      p2[2] = y;
     if (run(p2) == 19690720) {
        cout << x*100+y << "\n";
     }
    }
  }

  return 0;
}

File: adventofcode/2019/d2/d2.rs
fn run(program: Vec<usize>) -> usize {
  let mut arr = program.clone();
  let mut i = 0;
  loop {
    let op = arr[i];
    if op == 99 {
      return arr[0];
    }
    let a = arr[i + 1];
    let b = arr[i + 2];
    let to = arr[i + 3];
    if op == 1 {
      arr[to] = arr[a] + arr[b]
    } else if op == 2 {
      arr[to] = arr[a] * arr[b]
    }
    i += 4
  }
}

fn main() {
  let program: Vec<usize> = include_str!("./input.txt")
    .split(",")
    .map(|i| i.parse::<usize>().unwrap())
    .collect();

  // p1
  let mut p1 = program.clone();
  p1[1] = 12;
  p1[2] = 2;
  println!("{}", run(p1));

  // p2
  for x in 0..100 {
    for y in 0..100 {
      let mut p2 = program.clone();
      p2[1] = x;
      p2[2] = y;
      if run(p2) == 19690720 {
        println!("{}", 100 * x + y)
      }
    }
  }
}
